R/publish.R

```diff
@@ -8,23 +8,50 @@
 #' @param account Account name to publish to. Will default to any previously published
 #'  to account or any single account already associated with \code{server}.
 #' @param server Server to publish to (by default beta.rstudioconnect.com
-#' but any RStudio Connect server can be published to).
+#'  but any RStudio Connect server can be published to).
+#' @param render \code{TRUE} to render all formats prior to publishing (defaults to
+#'  \code{FALSE}, however this can be modified via the \code{bookdown.render_on_publish}
+#'  option). Note that this requires the use of a Makefile to provide the
+#'  implementaiton of rendering all formats.
 #'
 #' @export
 publish_book <- function(
-  name = NULL, account = NULL, server = 'beta.rstudioconnect.com'
+  name = NULL, account = NULL, server = NULL,
+  render = getOption("bookdown.render_on_publish", FALSE)
 ) {
 
   on.exit(opts$restore(), add = TRUE)
+
+  # render if requested (not we use the Makefile here rather than calling
+  # render_book because we need some mechanism that will have a-priori
+  # knowledge of all book formats)
+  if (isTRUE(render)) {
+    if (file.exists("Makefile")) {
+      result <- system2("make")
+      if (result != 0)
+        stop("Error ", result, " attempting to render book")
+    }
+    else
+      stop("Rendering before publishing requires a Makefile")
```

Comment:
I thought a while about this, and instead of `render_all(output_format = 'all')`, I think a better approach is to look for an R script, say, `_bookdown.R`, in which user can write arbitrary code to compile the book to any formats. This is because 1) there may be multiple types of HTML or PDF output, e.g. using both `html_book` and `gitbook`, or two PDFs with different templates, in which case the user needs to make sure these output files do not overwrite each other; 2) currently the MOBI output is not really created from an output format, but simply by converting EPUB; in other words, it cannot be generated by `render_book()`, but it is easy to create a MOBI e-book using `bookdown::kindlegen()`.